src/csharp/Grpc.Core/Internal/PlatformInvocationWin64.cs

```diff
@@ -0,0 +1,771 @@
+
+        ChannelState IPlatformInvocation.grpcsharp_channel_check_connectivity_state(ChannelSafeHandle channel, int tryToConnect)
+        {
+            return grpcsharp_channel_check_connectivity_state(channel, tryToConnect);
+        }
+
+        void IPlatformInvocation.grpcsharp_channel_watch_connectivity_state(ChannelSafeHandle channel,
+            ChannelState lastObservedState, Timespec deadline, CompletionQueueSafeHandle cq, BatchContextSafeHandle ctx)
+        {
+            grpcsharp_channel_watch_connectivity_state(channel, lastObservedState, deadline, cq, ctx);
+        }
+
+        CStringSafeHandle IPlatformInvocation.grpcsharp_channel_get_target(ChannelSafeHandle call)
+        {
+            return grpcsharp_channel_get_target(call);
+        }
+
+        void IPlatformInvocation.grpcsharp_channel_destroy(IntPtr channel)
+        {
+            grpcsharp_channel_destroy(channel);
+        }
+
+
+        // CompletionQueueEvent
+
+        [DllImport("grpc_csharp_ext-x64.dll")]
+        static extern int grpcsharp_sizeof_grpc_event();
+
+        int IPlatformInvocation.grpcsharp_sizeof_grpc_event()
+        {
+            return grpcsharp_sizeof_grpc_event();
+        }
+
+
+        // CompletionQueueSafeHandle
+
+        [DllImport("grpc_csharp_ext-x64.dll")]
+        static extern CompletionQueueSafeHandle grpcsharp_completion_queue_create();
+
+        [DllImport("grpc_csharp_ext-x64.dll")]
+        static extern void grpcsharp_completion_queue_shutdown(CompletionQueueSafeHandle cq);
+
+        [DllImport("grpc_csharp_ext-x64.dll")]
+        static extern CompletionQueueEvent grpcsharp_completion_queue_next(CompletionQueueSafeHandle cq);
+
+        [DllImport("grpc_csharp_ext-x64.dll")]
+        static extern CompletionQueueEvent grpcsharp_completion_queue_pluck(CompletionQueueSafeHandle cq, IntPtr tag);
+
+        [DllImport("grpc_csharp_ext-x64.dll")]
+        static extern void grpcsharp_completion_queue_destroy(IntPtr cq);
+
+        CompletionQueueSafeHandle IPlatformInvocation.grpcsharp_completion_queue_create()
+        {
+            return grpcsharp_completion_queue_create();
+        }
+
+        void IPlatformInvocation.grpcsharp_completion_queue_shutdown(CompletionQueueSafeHandle cq)
+        {
+            grpcsharp_completion_queue_shutdown(cq);
+        }
+        CompletionQueueEvent IPlatformInvocation.grpcsharp_completion_queue_next(CompletionQueueSafeHandle cq)
+        {
+            return grpcsharp_completion_queue_next(cq);
+        }
+
+        CompletionQueueEvent IPlatformInvocation.grpcsharp_completion_queue_pluck(CompletionQueueSafeHandle cq,
+            IntPtr tag)
+        {
+            return grpcsharp_completion_queue_pluck(cq, tag);
+        }
+
+        void IPlatformInvocation.grpcsharp_completion_queue_destroy(IntPtr cq)
+        {
+            grpcsharp_completion_queue_destroy(cq);
+        }
+
+
+
+        // CStringSafeHandle 
+
+        [DllImport("grpc_csharp_ext-x64.dll")]
```

Comment:
That would result in less nearly-duplicate code, and you could reduce the number of p/invoke calls this way, although because you still need to inspect the run-time environment, there will still be a need for an abstraction for the different targets.  It's not clear to me if that would be cleaner because DllImport takes care of path resolution (/lib:/usr/lib:..) on linux, etc.. and also takes care of freeing the library, which is internal code that you would need to reproduce.  I wouldn't recommend this because of the move to DNX/CoreCLR.  They seem to be in favor of DllMap, which is a monoism: https://github.com/dotnet/coreclr/issues/930, although that topic spurred a rather large thread which is still unresolved.